File: cpp/443_压缩字符串_medium.cpp
class Solution {
public:
    int compress(vector<char>& chars) {
        int n = chars.size();
        int write = 0, left = 0;
        for (int read = 0; read < n; read++) {
            if (read == n-1 || chars[read] != chars[read+1]) {
                chars[write++] = chars[read];
                int length = read - left + 1;
                if (length > 1) {
                    int tmp = write;
                    while (length > 0) {
                        chars[write++] = '0' + length%10;
                        length /= 10;
                    }
                    reverse(&chars[tmp], &chars[write]);
                }
                left = read + 1;
            }
        }
        return write;
    }
};

File: python/443_压缩字符串_medium.py
class Solution:
    def compress(self, chars: List[str]) -> int:
        def reverse(left, right):
            while left < right:
                chars[left], chars[right] = chars[right], chars[left]
                left += 1
                right -= 1

        n = len(chars)
        write, left = 0, 0
        for read in range(n):
            if read == n-1 or chars[read] != chars[read+1]:
                chars[write] = chars[read]
                write += 1
                length = read - left + 1
                if length > 1:
                    tmp = write
                    while length > 0:
                        chars[write] = str(length % 10)
                        write += 1
                        length //= 10
                    reverse(tmp, write-1)
                left = read + 1
        return write
